File: lib/domain/cultivation_plan/gateways/adjust_weather_prediction_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      # plan_allocation_adjust 用: WeatherPredictionInteractor 相当の予測サービスを生成する。
      class AdjustWeatherPredictionGateway
        # @param weather_location [Domain::WeatherData::Contracts::WeatherLocationPredictionInput]
        # @param farm [Domain::WeatherData::Contracts::FarmWeatherPredictionInput, nil]
        # @return [Object] #get_existing_prediction, #predict_for_cultivation_plan を実装
        def prediction_service(weather_location:, farm:)
          raise NotImplementedError, "Subclasses must implement prediction_service"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/adjust_weather_prediction_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::AdjustWeatherPredictionGateway`

use crate::weather_data::dtos::{CultivationPlanWeather, WeatherLocation};
use serde_json::Value;
use time::Date;

/// Prediction service: `get_existing_prediction` / `predict_for_cultivation_plan`.
pub trait WeatherPredictionService: Send + Sync {
    fn get_existing_prediction(
        &self,
        target_end_date: Date,
        cultivation_plan_weather: &CultivationPlanWeather,
    ) -> Option<Value>;

    fn predict_for_cultivation_plan(
        &self,
        cultivation_plan_weather: &CultivationPlanWeather,
        target_end_date: Option<Date>,
    ) -> Result<Value, Box<dyn std::error::Error + Send + Sync>>;
}

pub trait AdjustWeatherPredictionGateway: Send + Sync {
    fn prediction_service(
        &self,
        weather_location: &WeatherLocation,
    ) -> Result<Box<dyn WeatherPredictionService>, Box<dyn std::error::Error + Send + Sync>>;
}
